### ISSUE TITLE
Versioning should be semantic versioning compatible, doesn't seem to be

### ISSUE THREAD

**Issue by skylerbunny** [tootsuite/mastodon#8297]
(I realize it's too late to do anything about releases that are already out, but for future...)

See https://semver.org/.

Mastodon is versioning in such a way that it is difficult for an administrator of an instance to understand what is going into a version based on the version number, and therefore, to know what upgrade steps would be necessary to go from one version to another.

> Given a version number MAJOR.MINOR.PATCH, increment the:
> 
> MAJOR version when you make incompatible API changes,
> MINOR version when you add functionality in a backwards-compatible manner, and
> PATCH version when you make backwards-compatible bug fixes.
> 

At a minimum, I can see that version 2.4.3 states outright that it contains non-backwards compatible changes, including database migrations: ( https://github.com/tootsuite/mastodon/releases/tag/v2.4.3 )

> This release includes database migrations, that means you need to run RAILS_ENV=production bundle exec rails db:migrate (in Docker: docker-compose run --rm web rails db:migrate).

Releases with changes like this should come out as a M.m version, not an M.m.p version.

(If these releases are coming out with non-backwards compatible API changes, it should even be a new MAJOR version!)

* * * *

- [X] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by Gargron** [comment]
Semver refers to APIs. In Mastodon's case the existing APIs are the REST API, Streaming API, and federation-related (ActivityPub/OStatus) APIs. Major versions are those that introduce breaking changes in those APIs. For example, Mastodon 2.0 stopped federation of private messages over OStatus by switching to ActivityPub. We keep such breaking changes to a minimum, so there has not been a Mastodon 3.0 so far.

New functionality is introduced in minor patches, e.g. 2.1, 2.2, 2.3, 2.4, however, the line is very blurry between "new feature" and "polished UI" etc, so a lot of patch releases include more than just simple bug fixes. Semver is designed for programming libraries and packages, while Mastodon is end-user-facing software, so it's hard to follow it perfectly.

Database schema changes are internal, so do not have an effect on semver in my opinion.

**Comment by nightpool** [comment]
Gargron covered most of waht I was going to say, but semver doesn't make
sense for mastodon to begin with because it's not primarily consume by
other software—it's consumed by humans. There's no way for app developers
to "import mastodon 1.0"—if a server upgrades, they have to follow or stop
talking to that server.

Another note I wanted to add is that the REST and streaming APIs mastodon
implements are already versioned separately, using /v1/ and /v2/ routes. So
backwards incompatible changes are already accounted for in the software
itself, and (most of the time) don't require updating the major version
number

**Comment by skylerbunny** [comment]
The reason I brought this up is that humans _do_ need to upgrade Mastodon, and it would be useful to know either that a given major-minor-patch version should necessitate certain steps in performing said upgrade - or that you can easily roll back from one version to another if something goes wrong.

Alternatively, if the hidden message in https://github.com/tootsuite/documentation/blob/master/Running-Mastodon/Updating-Mastodon-Guide.md is that _all_ versions regardless of whether they "look like" MMp to someone familiar with semantic versioning should still do the same steps.

The specific case is, if I'm upgrading from 2.0.0 to 2.4.3, can I do it in one fell swoop reliably? The upgrade guide implies it because it never mentions step upgrades, but that's not an explicit statement. (If it doesn't, what version steps are needed? This is what brought me to this point.)
